File: app/src/main/java/com/codebosses/comical/activities/SignUpActivity.java
package com.codebosses.comical.activities;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.databinding.DataBindingUtil;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.view.View;

import com.codebosses.comical.R;
import com.codebosses.comical.common.Constants;
import com.codebosses.comical.databinding.ActivitySignUpBinding;
import com.codebosses.comical.endpoints.EndpointKeys;
import com.codebosses.comical.pojo.User;
import com.codebosses.comical.utils.L;
import com.codebosses.comical.utils.PrefUtils;
import com.codebosses.comical.utils.ValidUtils;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.AuthResult;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.firestore.FirebaseFirestore;
import com.ontbee.legacyforks.cn.pedant.SweetAlert.SweetAlertDialog;

public class SignUpActivity extends AppCompatActivity {

    //    Android fields....
    private ActivitySignUpBinding signUpBinding;
    private Context context;
    private SweetAlertDialog sweetAlertDialog;

    //    Firebase fields....
    private FirebaseAuth firebaseAuth;
    private FirebaseFirestore firebaseFirestore;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        signUpBinding = DataBindingUtil.setContentView(this, R.layout.activity_sign_up);

        context = this;

//        Android fields initialization....
        sweetAlertDialog = new SweetAlertDialog(context);
        sweetAlertDialog.changeAlertType(SweetAlertDialog.PROGRESS_TYPE);
        sweetAlertDialog.setCancelable(false);
        sweetAlertDialog.setTitleText("Creating account...");

//        Firebase fields initialization....
        firebaseAuth = FirebaseAuth.getInstance();
        firebaseFirestore = FirebaseFirestore.getInstance();

        signUpBinding.setClickHandler(new SignUpClickHandler());
    }

    private void validateUser() {
        String name = signUpBinding.editTextNameSignUp.getText().toString();
        String email = signUpBinding.editTextEmailSignUp.getText().toString();
        String password = signUpBinding.editTextPasswordSignUp.getText().toString();
        if (ValidUtils.validateEditTexts(signUpBinding.editTextNameSignUp, signUpBinding.editTextEmailSignUp, signUpBinding.editTextPasswordSignUp)) {
            if (ValidUtils.validateEmail(signUpBinding.editTextEmailSignUp)) {
                if (ValidUtils.validatePassword(password)) {
                    createUser(name, email, password);
                } else {
                    L.showToast(context, getResources().getString(R.string.invalid_password));
                }
            } else {
                L.showToast(context, getResources().getString(R.string.invalid_email_address));
            }
        } else {
            L.showToast(context, getResources().getString(R.string.please_fill_all_fields));
        }
    }

    private void createUser(String name, String email, String passsword) {
        sweetAlertDialog.show();
        firebaseAuth.createUserWithEmailAndPassword(email, passsword)
                .addOnCompleteListener(SignUpActivity.this, new OnCompleteListener<AuthResult>() {
                    @Override
                    public void onComplete(@NonNull Task<AuthResult> task) {
                        if (task.isSuccessful()) {
                            User user = new User(task.getResult().getUser().getUid(), name, email, "", "", "", Constants.EMAIL);
                            firebaseFirestore.collection(EndpointKeys.USERS)
                                    .document(task.getResult().getUser().getUid())
                                    .set(user)
                                    .addOnCompleteListener(SignUpActivity.this, new OnCompleteListener<Void>() {
                                        @Override
                                        public void onComplete(@NonNull Task<Void> task) {
                                            sweetAlertDialog.dismiss();
                                            if (task.isSuccessful()) {
                                                PrefUtils.saveToPrefs(context, EndpointKeys.USER_NAME, user.getUserName());
                                                PrefUtils.saveToPrefs(context, EndpointKeys.USER_EMAIL, user.getUserEmail());
                                                PrefUtils.saveToPrefs(context, EndpointKeys.USER_PASSWORD, passsword);
                                                Intent intent = new Intent(context, MainActivity.class);
                                                intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
                                                startActivity(intent);
                                                finish();
                                            } else {
                                                L.showToast(context, task.getException().getMessage());
                                            }
                                        }
                                    });
                        } else {
                            sweetAlertDialog.dismiss();
                            L.showToast(context, task.getException().getMessage());
                        }
                    }
                });
    }

    public class SignUpClickHandler {

        public void onSignUpClick(View view) {
            validateUser();
        }

        public void onBackClick(View view) {
            onBackPressed();
        }

    }
}

File: app/src/main/java/com/codebosses/comical/ui/registration/signup/SignUpActivity.kt
package com.codebosses.comical.ui.registration.signup

import android.os.Bundle
import android.view.View
import com.codebosses.comical.R
import com.codebosses.comical.repository.model.user.UserResult
import com.codebosses.comical.ui.BaseActivity
import com.codebosses.comical.ui.main.MainActivity
import com.codebosses.comical.ui.registration.login.LoginViewModel
import com.codebosses.comical.utils.PrefUtils
import com.codebosses.comical.utils.StringUtils
import com.codebosses.comical.utils.ToastUtil
import com.codebosses.comical.utils.extensions.getViewModel
import com.codebosses.comical.utils.extensions.observe
import com.codebosses.comical.utils.extensions.startActivityNewTask
import com.ontbee.legacyforks.cn.pedant.SweetAlert.SweetAlertDialog
import kotlinx.android.synthetic.main.activity_sign_up.*

class SignUpActivity : BaseActivity(), View.OnClickListener {


    //    View model field....
    private val signUpViewModel by lazy {
        getViewModel<SignUpViewModel>()
    }

    //    Android fields....
    private lateinit var sweetAlertDialog: SweetAlertDialog

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_sign_up)

        //        Dialog initialization....
        sweetAlertDialog = SweetAlertDialog((this))
        sweetAlertDialog.changeAlertType(SweetAlertDialog.PROGRESS_TYPE)
        sweetAlertDialog.titleText = "Please wait..."
        sweetAlertDialog.setCancelable(false)

//        Listeners....
        textViewEnterWithLogin.setOnClickListener(this)
        buttonSignUp.setOnClickListener(this)
    }

    override fun onClick(view: View?) {
        when (view!!.id) {
            R.id.textViewEnterWithLogin -> {
                finish()
            }
            R.id.buttonSignUp -> {
                val name = editTextNameSignUp.text.toString()
                val email = editTextEmailSignUp.text.toString()
                val password = editTextPasswordSignUp.text.toString()
                createUserWithEmailAndPassword(name, email, password)
            }
        }
    }

    private fun createUserWithEmailAndPassword(name: String, email: String, password: String) {
        if (validateData(name, email, password)) {
            signUpViewModel.createUserWithEmailAndPassword(name, email, password, "", 1, PrefUtils.deviceToken)
                    .observe(this) {
                        when {
                            it.status.isLoading() -> {
                                sweetAlertDialog.show()
                            }
                            it.status.isSuccessful() -> {
                                sweetAlertDialog.dismiss()
                                saveDataInPref(it.data!!.data)
                            }
                            it.status.isError() -> {
                                sweetAlertDialog.dismiss()
                                ToastUtil.showCustomToast(this, it.errorMessage!!)
                            }
                        }
                    }
        }
    }

    private fun validateData(name: String, email: String, password: String): Boolean {
        if (name.isNotEmpty() && email.isNotEmpty() && password.isNotEmpty()) {
            if (StringUtils.isValidEmail(email)) {
                if (password.length > 5) {
                    return true
                } else {
                    ToastUtil.showCustomToast(this, resources.getString(R.string.invalid_password))
                    return false
                }
            } else {
                ToastUtil.showCustomToast(this, resources.getString(R.string.invalid_email_address))
                return false
            }
        } else {
            ToastUtil.showCustomToast(this, resources.getString(R.string.please_fill_all_fields))
            return false
        }
    }

    private fun saveDataInPref(userResult: UserResult) {
        PrefUtils.isUserLoggedIn = true
        PrefUtils.userId = userResult.user_id
        PrefUtils.userName = userResult.user_name
        PrefUtils.userEmail = userResult.user_email
        PrefUtils.profileImageUrl = userResult.profile_image_url
        startActivityNewTask(MainActivity::class.java)
    }
}
